packages/gtmapi/lmsrvcore/auth/identity.py

```
@@ -1,29 +1,56 @@
 
 
-def parse_token(auth_header: str) -> str:
-    """Method to extract the bearer token from the authorization header
+def tokens_from_headers(headers: dict) -> Tuple[Optional[str], Optional[str]]:
+    """Method to retrieve the access and id tokens from the request headers, if available."""
+    access_token = None
+    id_token = None
+    if "Authorization" in headers:
+        try:
+            _, access_token = headers['Authorization'].split("Bearer ")
+        except ValueError:
+            raise AuthenticationError("Could not parse JWT from Authorization Header. Should be `Bearer XXX`", 401)
+
+    if "Identity" in headers:
+        id_token = headers['Identity']
+
+    return access_token, id_token
+
+
+def tokens_from_request_context(tokens_required=False) -> Tuple[Optional[str], Optional[str]]:
+    """Method to return tokens stored in the flask request context. If `tokens_required` is True, an AuthenticationError
+    exception will be raised if tokens are not present.
 
     Args:
-        auth_header(str): The Authorization header
+        tokens_required: boolean indicating if tokens must exist
 
-    Returns:
-        str
     """
-    if "Bearer" in auth_header:
-        _, token = auth_header.split("Bearer ")
-        if not token:
-            raise AuthenticationError("Could not parse JWT from Authorization Header. Should be `Bearer XXX`", 401)
-    else:
-        raise AuthenticationError("Could not parse JWT from Authorization Header. Should be `Bearer XXX`", 401)
-
-    return token
+    access_token = flask.g.get('access_token', None)
+    id_token = flask.g.get('id_token', None)
+
+    if tokens_required:
+        if not access_token or not id_token:
+            raise AuthenticationError("This operation requires a valid session. Please log in.", 401)
+
+    # If the token is an anonymous token, set tokens to None so future requests and git operations
+    # try to run without authentication.
+    parts = access_token.split('.')
```

The logic here seems off. Line 44 above implies that in regular auth, we expect both access_token and id_token. This line and 52 below implies that we expect an access_token in anonymous auth also.

So, is something always broken if access_token is missing? If so, this function should return an error, right?